platform/ui/src/contextProviders/ViewportGridProvider.tsx

```diff
@@ -59,135 +99,138 @@ export function ViewportGridProvider({ children, service }) {
-        // into the viewport object at the given index
-        viewports[viewportIndex] = {
-          ...viewports[viewportIndex],
+        const viewports = state.viewports.slice();
+
+        const newView = {
+          ...viewport,
           displaySetInstanceUIDs,
           viewportOptions,
           displaySetOptions,
           viewportLabel: viewportLabels[viewportIndex],
         };
+        viewportOptions.presentationId = getPresentationId(newView, viewports);
+
+        // Make sure we assign a viewport id
+        if (newView.viewportId || viewportOptions.viewportId) {
+          newView.viewportId = newView.viewportOptions.viewportId =
+            newView.viewportId || viewportOptions.viewportId;
+        } else {
+          newView = reuseViewport({}, newView, state.viewports);
+        }
+
+        viewports[viewportIndex] = newView;
 
-        return { ...state, ...{ viewports } };
+        return { ...state, viewports };
       }
       case 'SET_LAYOUT': {
         const {
           numCols,
           numRows,
           layoutOptions,
           layoutType = 'grid',
-          keepExtraViewports = false,
+          findOrCreateViewport,
         } = action.payload;
 
         // If empty viewportOptions, we use numRow and numCols to calculate number of viewports
-        const numPanes = layoutOptions.length || numRows * numCols;
-        const viewports = state.viewports.slice();
-        const activeViewportIndex =
-          state.activeViewportIndex >= numPanes ? 0 : state.activeViewportIndex;
-
-        while (viewports.length < numPanes) {
-          viewports.push({});
-        }
-
-        // Extra viewports are kept when the grid layout is changed in the UI
-        // because the user populated those viewports and if the viewports were to
-        // return on screen their contents should be maintained.
-        if (!keepExtraViewports) {
-          while (viewports.length > numPanes) {
-            viewports.pop();
+        const hasOptions = layoutOptions?.length;
+        const viewports = [];
+
+        // Options is a cache of values allowing for findOrCreate to store
+        // information on progress, reset on every state update
+        const options = {};
+
+        let activeViewportIndex;
+        for (let row = 0; row < numRows; row++) {
+          for (let col = 0; col < numCols; col++) {
+            const pos = col + row * numCols;
+            const layoutOption = layoutOptions[pos];
+            const positionId = layoutOption?.positionId || `${col}-${row}`;
+            if ((hasOptions && pos < layoutOptions.length) || !hasOptions) {
+              if (
+                !activeViewportIndex ||
+                state.viewports[pos]?.positionId === positionId
+              ) {
+                activeViewportIndex = pos;
+              }
+              const viewport = findOrCreateViewport(pos, positionId, options);
```

Comment:
I do not see a `findOrCreateViewport` implementation that takes three arguments. I see two: one that takes five arguments and another that takes one. What am I missing?